### PR TITLE
Arm64 support in dockerfile.merlin

### LINKED ISSUE #834: [FEA] ARM support
# 🚀 Feature request

HugeCTR ARM support is currently blocked by the fact that we don't have a fitting base image.

## Motivation

- Grace is inbound
- Inquiries from customers present.

### PR DESCRIPTION
**Pull request by jlinford**
Resolves #834 to add support for Arm64 in the Merlin base image.  Additional changes are needed in other dockerfiles to broadly enable Arm64 support, but this is a start.

### CODE REVIEW

**Comment by jlinford** [comment]
I've successfully built on both arm64 and amd64 via `docker build - < docker/dockerfile.merlin`.

**Comment by karlhigley** [comment]
This LGTM but I want to get eyes on it from the people working on our container releases before pushing the big green button 😄

File: docker/dockerfile.merlin [hunk 1]
```diff
@@ -144,7 +164,17 @@ RUN rm -rf /repos
...
-FROM ${BASE_IMAGE} as base
+FROM ${BASE_IMAGE} as base-amd64
+ONBUILD COPY --chown=1000:1000 --from=triton /usr/lib/x86_64-linux-gnu/libdcgm.so.2 /usr/lib/x86_64-linux-gnu/libdcgm.so.2
+ONBUILD COPY --chown=1000:1000 --from=triton /usr/local/cuda-11.8/targets/x86_64-linux/lib/libcupti.so.11.8 /usr/local/cuda-11.8/targets/x86_64-linux/lib/libcupti.so.11.8
+ONBUILD COPY --chown=1000:1000 --from=triton /opt/tritonserver/backends/fil /opt/tritonserver/backends/fil/
+
+FROM ${BASE_IMAGE} as base-arm64
+ONBUILD COPY --chown=1000:1000 --from=triton /usr/lib/aarch64-linux-gnu/libdcgm.so.2 /usr/lib/aarch64-linux-gnu/libdcgm.so.2
+ONBUILD COPY --chown=1000:1000 --from=triton /usr/local/cuda-11.8/targets/sbsa-linux/lib/libcupti.so.11.8 /usr/local/cuda-11.8/targets/sbsa-linux/lib/libcupti.so.11.8
+RUN echo "Skipping copy of /opt/tritonserver/backends/fil. (Why does this fail on arm64?)"
```
Comment (jperez999) [inline docker/dockerfile.merlin:175]: Cant these two sets of instructions be condensed? They are running from the same base image. Or am I misunderstanding what is happening here?

File: docker/dockerfile.merlin [hunk 2]
```diff
@@ -10,6 +10,14 @@ FROM ${FULL_IMAGE} as triton
 FROM ${DLFW_IMAGE} as dlfw
 FROM ${BASE_IMAGE} as build
 
+FROM ${BASE_IMAGE} as build-amd64
+ONBUILD COPY --chown=1000:1000 --from=triton /opt/tritonserver/backends/fil /opt/tritonserver/backends/fil/
+
+FROM ${BASE_IMAGE} as build-arm64
+RUN echo "Skipping copy of /opt/tritonserver/backends/fil. (Why does this fail on arm64?)"
```
Comment (jperez999) [inline docker/dockerfile.merlin:17]: Why are these calls done separately? The base image will be the same throughout the build right?

File: docker/dockerfile.merlin [hunk 3]
```diff
@@ -88,17 +100,26 @@ RUN ln -s /usr/bin/python3 /usr/bin/python
...
 # A fix has already been merged but not yet released:
 # https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7859
-# 2023-02-22: pynvml==11.5.0 is currently incompatible with our version of dask/distributed
-RUN pip install --no-cache-dir --upgrade pip; pip install --no-cache-dir "cmake<3.25.0" ninja scikit-build pandas==1.3.5 \
-                cupy-cuda117 nvidia-pyindex pybind11 pytest \
+RUN pip install --no-cache-dir --upgrade pip; pip install --no-cache-dir "cmake<3.25.0" ninja scikit-build pandas==1.3.5
+
+# cupy-cuda wheels come from a different URL on aarch64
+RUN if [ $(uname -m) == "aarch64" ] ; then \
+        pip install cupy-cuda115 -f https://pip.cupy.dev/aarch64/ ; \
```
Comment (jperez999) [inline docker/dockerfile.merlin:107]: This logic implies that we are working with cuda 11. I think that in march we are moving to cuda 12. Are we going to need to support cuda 11.x for ARM support when that shift happens?

File: docker/dockerfile.merlin [hunk 4]
```diff
@@ -252,13 +280,21 @@ ENV JUPYTER_CONFIG_DIR=/tmp/.jupyter
...
-ARG DL_VER
+ARG MERLIN_VER=main
+ARG CORE_VER=main
+ARG MODELS_VER=main
+ARG NVTAB_VER=main
+ARG SYSTEMS_VER=main
+ARG TF4REC_VER=main
+ARG DL_VER=main
+
+ENV MERLIN_VER=${MERLIN_VER}
```
Comment (jperez999) [inline docker/dockerfile.merlin:291]: What is the reason for storing these values in the container? We use them for pulling in the code in the following instructions. Are we changing something that will leverage this?

File: docker/dockerfile.merlin (hunk 2 above)
Comment (jlinford) [reply docker/dockerfile.merlin:17]: They must be separated, not combined.  Lines 13-14 are for amd64 (i.e. x86_64) while lines 16-17 are for arm64 (i.e aarch64).  This is one way to do a conditional copy in Docker where different file paths are copied depending on the target architecture.

File: docker/dockerfile.merlin (hunk 3 above)
Comment (jlinford) [reply docker/dockerfile.merlin:107]: I think that's a discussion for a different PR.  This PR only adds arm64 support to the existing Dockerfile.  If we need to change the CUDA version, that's a different PR and completely independent of adding arm64 support.  Both CUDA 11 and CUDA 12 work fine on arm64, so we could go either way, but arm64 should match x86 regardless.

File: docker/dockerfile.merlin (hunk 1 above)
Comment (jlinford) [reply docker/dockerfile.merlin:175]: Same as above, they must be separated, not combined. Lines 167-170 are for amd64 (i.e. x86_64) while lines 172-175 are for arm64 (i.e aarch64). Note how the amd64 filepaths contain `x86_64` while the arm64 filepaths contain `sbsa`.  This is one way to do a conditional copy in Docker where different file paths are copied depending on the target architecture.

File: docker/dockerfile.merlin (hunk 4 above)
Comment (jlinford) [reply docker/dockerfile.merlin:291]: They were already stored in the container at the top of the Dockerfile on Lines 14-22.  They needed to be replicated here (well, somewhere after the `FROM ${BASE_IMAGE} as base` command) so the subsequent commands like `git clone --branch ${MERLIN_VER}` would work.  Without specifying `=main`, the default is empty and that `git clone` command fails when building via commands like `docker build - < docker/dockerfile.merlin` where the argument value isn't explicitly set.

**Comment by jperez999** [comment]
@jlinford does the build command have to change for all builds? Like to build the non-arm docker image do I need to use `docker build - < docker/dockerfile.merlin` or does `docker build -t <tag name> -f docker/dockerfile.merlin .` ? If so, because we are in release right now I am going to hold off until that is complete before adding this because we will need to change CI commands in the container build process.

**Comment by jlinford** [comment]
The build command should not change, either on arm64 or x86_64.  On arm64, it should be possible to use the exact same build command you currently use on x86_64, and get an arm64 image as a result.  If that doesn't work, I'd consider it a bug and would want to fix it.

**Comment by jlinford** [comment]
Apologies, I think I accidentally removed @oliverholworthy and @edknv from the reviewers, but somehow can't add them back.  Happy to have their review!

**Review by jperez999 — APPROVED** [review]

**Comment by jperez999** [comment]
I approved... but we will wait to merge. Sorry I should have done that from the beginning.

**Pull request merged**